src/Miki.Framework.Discord/Services/DiscordHostedService.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Miki.Discord.Common;
+using Miki.Framework.Discord.Factories;
+using Miki.Framework.Hosting;
+using Miki.Framework.Models;
+using Miki.Logging;
+
+namespace Miki.Framework.Discord.Services
+{
+    public class DiscordHostedService : IHostedService, IDisposable
+    {
+        private readonly IServiceProvider serviceProvider;
+        private readonly MessageDelegate invoke;
+        private readonly IDiscordClientFactory discordClientFactory;
+        private IDiscordClient discordClient;
+
+        public DiscordHostedService(
+            IBotApplicationBuilderFactory factory,
+            IServiceProvider serviceProvider,
+            IDiscordClientFactory discordClientFactory)
+        {
+            this.serviceProvider = serviceProvider;
+            this.discordClientFactory = discordClientFactory;
+            invoke = factory.CreateBuilder().Build();
+        }
+
+        private async Task HandleMessageAsync(IDiscordMessage message)
+        {
+            using var scope = serviceProvider.CreateScope();
+            using var context = new ContextObject(scope.ServiceProvider, new DiscordMessage(message));
+
+            context.SetContext("DiscordClient", discordClient);
+            
+            try
+            {
+                await invoke(context);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e);
```

Comment:
Is this before the pipeline?

Reply:
No, this is after the pipeline. This is called when an exception occurred in the middelware.

Reply:
Can we add an event then when an exception occurs to fetch the current context state and exception?